<launch>
  <node pkg="jsk_topic_tools" type="relay" name="input_relay">
    <remap from="~input" to="/multisense_local/resize_1_4/points" />
  </node>
  <node pkg="jsk_pcl_ros" type="organized_multi_plane_segmentation" name="multi_plane_estimate">
    <remap from="~input" to="input_relay/output"/>
    <rosparam>
      max_curvature: 0.01
      estimate_normal: true
    </rosparam>
  </node>
  <node pkg="nodelet" type="nodelet"
        name="plane_extraction"
        args="standalone jsk_pcl/MultiPlaneExtraction"
        output="screen">
    <remap from="~input" to="input_relay/output" />
    <remap from="~indices" to="/multi_plane_estimate/output_refined" />
    <remap from="~input_polygons" to="/multi_plane_estimate/output_refined_polygon" />
    <remap from="~input_coefficients" to="/multi_plane_estimate/output_refined_coefficients" />
    <rosparam>
      use_sensor_frame: true
      sensor_frame: head_root
    </rosparam>
  </node>

  <node pkg="nodelet" type="nodelet" name="euclidean_clustering"
        args="standalone jsk_pcl/EuclideanClustering " output="screen">
    <remap from="~input" to="/plane_extraction/output" />
    <rosparam>
      tolerance: 0.02
      min_size: 100
    </rosparam>
  </node>

  <node pkg="nodelet" type="nodelet"
        name="throttle_segmentation"
        args="standalone jsk_topic_tools/LightweightThrottle"
        output="screen">
    <remap from="~input" to="euclidean_clustering/output" />
    <remap from="~output" to="euclidean_clustering/output_throttle" />
  </node>

  
  <node pkg="nodelet" type="nodelet"
        name="segmentation_decomposer"
        args="standalone jsk_pcl/ClusterPointIndicesDecomposer"
        output="screen">
    <remap from="~input" to="/plane_extraction/output" />
    <remap from="~target" to="/euclidean_clustering/output_throttle" />
    <remap from="~align_planes" to="/multi_plane_estimate/output_refined_polygon" />
    <remap from="~align_planes_coefficients"
           to="/multi_plane_estimate/output_refined_coefficients" />
    <rosparam>
      align_boxes: true
      use_pca: true
      publish_clouds: false
      publish_tf: false
    </rosparam>
  </node>

  <node pkg="jsk_interactive_marker"
        type="bounding_box_marker"
        name="bounding_box_marker"
        output="screen"
        >
    <remap from="~bounding_box_array" to="segmentation_decomposer/boxes" />
  </node>
  <node pkg="nodelet" type="nodelet"
        name="selected_cloud"
        args="standalone jsk_pcl/SelectedClusterPublisher"
        output="screen">
    <remap from="~input" to="/plane_extraction/output" />
    <remap from="~indices" to="/euclidean_clustering/output" />
    <remap from="~selected_index" to="/bounding_box_marker/selected_index" />
    <remap from="~output" to="/selected_pointcloud" />
  </node>

  <node pkg="nodelet" type="nodelet"
        name="octree_change_detector"
        args="standalone jsk_pcl/OctreeChangePublisher">
    <remap from="~input" to="voxelgrid/output" />
    <rosparam>
      resolution: 0.1
      noise_filter: 8
    </rosparam>
  </node>

  <node pkg="nodelet" type="nodelet"
        name="voxelgrid"
        args="standalone jsk_pcl/OctreeVoxelGrid"
        output="screen" clear_params="true">
    <remap from="~input" to="input_relay/output" />
    <param name="resolution" value="0.01" />
    <rosparam>
      point_type: xyzrgb
      marker_color_alpha: 0.5
    </rosparam>
  </node>

    <node pkg="nodelet" type="nodelet"
          name="model_voxelgrid"
        args="standalone jsk_pcl/OctreeVoxelGrid"
        output="screen" clear_params="true">
    <remap from="~input" to="/selected_pointcloud" />
    <param name="resolution" value="0.01" />
    <rosparam>
      point_type: xyzrgb
      marker_color_alpha: 0.5
    </rosparam>
  </node>


  
  <node pkg="nodelet" type="nodelet"
        name="particle_filter_tracker"
        args="standalone jsk_pcl/ParticleFilterTracking"
        output="screen" clear_params="true">
    <remap from="~input" to="voxelgrid/output" />
    <remap from="~input_change" to="octree_change_detector/octree_change_result" />
    <remap from="~renew_model" to="model_voxelgrid/output" />
    <rosparam>
      max_particle_num: 1000
      use_change_detection: true
      default_step_covariance_x: 0.00004
      default_step_covariance_y: 0.00004
      default_step_covariance_z: 0.00004
    </rosparam>
  </node>
  
</launch>
